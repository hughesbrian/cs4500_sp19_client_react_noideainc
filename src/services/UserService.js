export default class UserService {
    static instance = null;

    static getInstance() {
        if (UserService.instance === null) {
            UserService.instance = new UserService()
        }
        return this.instance
    }

    findUserById = userId =>
        fetch("https://cs4500-sp19-noideainc.herokuapp.com/api/users/${userId}")
            .then(response => response.json())
    findAllUsers = () =>
        fetch("https://cs4500-sp19-noideainc.herokuapp.com/api/users")
            .then(response => response.json())
}