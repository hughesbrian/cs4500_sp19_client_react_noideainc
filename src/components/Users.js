import React from 'react'
import UserService from '../services/UserService'
import {BrowserRouter as Router, Link, Route} from 'react-router-dom'

class Users extends React.Component {
    constructor(props) {
        super(props);
        this.userService = UserService.getInstance();
        this.state = {
            editingUser: {
                username: '',
                password: '',
                firstName: '',
                lastName: ''
            },
            users: []
        }
    }

    componentDidMount() {
        this.userService
            .findAllUsers()
            .then(users =>
                this.setState({
                    users: users
                })
            )
    }

    createUser = (user) =>
        this.userService
            .createUser(user)
            .then(this.props.history.push("/admin/users/"));

    saveUser = function (user) {
        if (user.id) {
            this.updateUser(user);
        } else {
            this.createUser(user);
        }
    };

    updateUser = (user) =>
        this.userService
            .updateUser(user)
            .then(this.props.history.push("/admin/users/"));

    deleteUser = (user) =>
        this.userService
            .deleteUser(user)
            .then(this.props.history.push("/admin/users/"));

    render() {
        return (
            <div>
                <h3>Users</h3>

                <table className="table">
                    <tbody>
                    <tr>
                        <td><input onChange={(e) =>
                            this.setState({
                                editingUser: {
                                    id: this.state.editingUser.id,
                                    username: e.target.value,
                                    password: this.state.editingUser.password,
                                    firstName: this.state.editingUser.firstName,
                                    lastName: this.state.editingUser.lastName
                                }
                            })} className="form-control"
                                   value={this.state.editingUser.username}/></td>
                        <td><input onChange={(e) =>
                            this.setState({
                                editingUser: {
                                    id: this.state.editingUser.id,
                                    username: this.state.editingUser.username,
                                    password: this.state.editingUser.password,
                                    firstName: e.target.value,
                                    lastName: this.state.editingUser.lastName
                                }
                            })} className="form-control"
                                   value={this.state.editingUser.firstName}/></td>
                        <td><input onChange={(e) =>
                            this.setState({
                                editingUser: {
                                    id: this.state.editingUser.id,
                                    username: this.state.editingUser.username,
                                    password: this.state.editingUser.password,
                                    firstName: this.state.editingUser.firstName,
                                    lastName: e.target.value
                                }
                            })} className="form-control"
                                   value={this.state.editingUser.lastName}/></td>
                        <td>
                            <button className="btn btn-success" onClick={() => {
                                this.saveUser(this.state.editingUser)
                            }}>Save
                            </button>
                            <button className="btn btn-warning" onClick={() => {
                                this.setState({
                                    editingUser: {
                                        username: '',
                                        password: '',
                                        firstName: '',
                                        lastName: ''
                                    }
                                });
                            }}>Clear
                            </button>
                        </td>
                    </tr>
                    {
                        this.state.users
                            .map(user =>
                                <tr key={user.id}>
                                    <td>{user.username}</td>
                                    <td>{user.firstName}</td>
                                    <td>{user.lastName}</td>
                                    <td>
                                        <button className="btn btn-primary" onClick={() => {
                                            this.setState({editingUser: user});
                                        }}>Edit
                                        </button>
                                        <button className="btn btn-danger" onClick={
                                            () => this.deleteUser(user)
                                        }>Delete
                                        </button>
                                    </td>
                                </tr>
                            )
                    }
                    </tbody>
                </table>
            </div>
        )
    }
}

export default Users