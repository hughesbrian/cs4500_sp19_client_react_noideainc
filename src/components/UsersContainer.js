import React from 'react';
import Users from './Users';
import UserService from "../services/UserService";

class UsersContainer extends React.Component {
    constructor(props) {
        super(props);
        this.userService = UserService.getInstance();
        this.state = {
            editingUser: {
                username: '',
                password: '',
                firstName: '',
                lastName: ''
            },
            users: []
        }
    }

    componentDidMount() {
        this.userService
            .findAllUsers()
            .then(users =>
                this.setState({
                    users: users
                })
            )
    }

    createUser = (user) =>
        this.userService
            .createUser(user)
            .then(this.props.history.push("/admin/users/"));

    saveUser = function (user, updateUser, createUser) {
        if (user.id) {
            updateUser(user);
        } else {
            createUser(user);
        }
    };

    setEditingUser = (new_attrs) =>
        this.setState(new_attrs);

    updateUser = (user) =>
        this.userService
            .updateUser(user)
            .then(this.props.history.push("/admin/users/"));

    deleteUser = (user) =>
        this.userService
            .deleteUser(user)
            .then(this.props.history.push("/admin/users/"));

    render = () => <Users users={this.state.users}
                          editingUser={this.state.editingUser}
                          setEditingUser={this.setEditingUser}
                          createUser={this.createUser}
                          updateUser={this.updateUser}
                          deleteUser={this.deleteUser}
                          saveUser={this.saveUser}/>;
}

export default UsersContainer