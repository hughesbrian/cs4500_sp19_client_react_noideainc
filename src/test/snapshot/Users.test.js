import React from 'react'
import Users from '../../components/Users'
import TestRenderer from 'react-test-renderer'
import users from '../MockData/Users.mock'
import UserService from '../../services/UserService'
import '../service/UserService.mock'
import UsersContainer from "../../components/UsersContainer";

const userService = UserService.getInstance();

test('Create and Delete Users', () => {
    let spidey = {id: 345, username: "pparker", password: "iamspiderman", firstName: "Peter", lastName: "Parker"};

    const saveUser = () => {
        users.push(spidey);

        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();

    };

    const deleteUser = userId => {
        users.splice(0, 1)

        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();

    };

    const testRenderer = TestRenderer.create(
        <Users users={users} editingUser={spidey} deleteUser={deleteUser} saveUser={saveUser}/>
    );
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();

    const testInstance = testRenderer.root;

    const createUserBtn = testInstance.findByProps({className: 'btn btn-success save-user-btn'});
    createUserBtn.props.onClick();
    createUserBtn.props.onClick();

    const deleteUserBtns = testInstance.findAllByProps({className: 'btn btn-danger delete-user-btn'});
    deleteUserBtns[1].props.onClick();
});

test('List of Users Renders Directly', () => {
    const testRenderer = TestRenderer.create(
        <Users users={users} editingUser={users[0]}/>
    );

    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();

    const testInstance = testRenderer.root;

    const usernameField = testInstance.findByProps({className: 'form-control username-edit-field'});
    const numUsers = testInstance.findAllByProps({className: 'btn btn-danger delete-user-btn'});

    expect(usernameField.props.value).toBe('nate');
    expect(numUsers.length).toBe(users.length);
});

test('List of Users Renders from Service', () => {
    userService.findAllUsers().then(users => {
        const testRenderer = TestRenderer.create(
            <Users users={users} editingUser={users[0]}/>
        );

        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot();

        const testInstance = testRenderer.root;

        const usernameField = testInstance.findByProps({className: 'form-control username-edit-field'});
        const numUsers = testInstance.findAllByProps({className: 'btn btn-danger delete-user-btn'});

        expect(usernameField.props.value).toBe('nate');
        expect(numUsers.length).toBe(users.length);
    });
});