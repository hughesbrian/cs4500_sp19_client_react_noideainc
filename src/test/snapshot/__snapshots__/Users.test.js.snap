// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create and Delete Users 1`] = `
<div>
  <h3>
    Users
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="form-control username-edit-field"
            onChange={[Function]}
            value="pparker"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Peter"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Parker"
          />
        </td>
        <td>
          <button
            className="btn btn-success save-user-btn"
            onClick={[Function]}
          >
            Save
          </button>
          <button
            className="btn btn-warning"
            onClick={[Function]}
          >
            Clear
          </button>
        </td>
      </tr>
      <tr>
        <td>
          nate
        </td>
        <td>
          Nathan
        </td>
        <td>
          Jones
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          sam
        </td>
        <td>
          Sam
        </td>
        <td>
          Atkins
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Create and Delete Users 2`] = `
<div>
  <h3>
    Users
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="form-control username-edit-field"
            onChange={[Function]}
            value="pparker"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Peter"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Parker"
          />
        </td>
        <td>
          <button
            className="btn btn-success save-user-btn"
            onClick={[Function]}
          >
            Save
          </button>
          <button
            className="btn btn-warning"
            onClick={[Function]}
          >
            Clear
          </button>
        </td>
      </tr>
      <tr>
        <td>
          nate
        </td>
        <td>
          Nathan
        </td>
        <td>
          Jones
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          sam
        </td>
        <td>
          Sam
        </td>
        <td>
          Atkins
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Create and Delete Users 3`] = `
<div>
  <h3>
    Users
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="form-control username-edit-field"
            onChange={[Function]}
            value="pparker"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Peter"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Parker"
          />
        </td>
        <td>
          <button
            className="btn btn-success save-user-btn"
            onClick={[Function]}
          >
            Save
          </button>
          <button
            className="btn btn-warning"
            onClick={[Function]}
          >
            Clear
          </button>
        </td>
      </tr>
      <tr>
        <td>
          nate
        </td>
        <td>
          Nathan
        </td>
        <td>
          Jones
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          sam
        </td>
        <td>
          Sam
        </td>
        <td>
          Atkins
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`List of Users Renders Directly 1`] = `
<div>
  <h3>
    Users
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="form-control username-edit-field"
            onChange={[Function]}
            value="sam"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Sam"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Atkins"
          />
        </td>
        <td>
          <button
            className="btn btn-success save-user-btn"
            onClick={[Function]}
          >
            Save
          </button>
          <button
            className="btn btn-warning"
            onClick={[Function]}
          >
            Clear
          </button>
        </td>
      </tr>
      <tr>
        <td>
          sam
        </td>
        <td>
          Sam
        </td>
        <td>
          Atkins
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          pparker
        </td>
        <td>
          Peter
        </td>
        <td>
          Parker
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`List of Users Renders from Service 1`] = `
<div>
  <h3>
    Users
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="form-control username-edit-field"
            onChange={[Function]}
            value="sam"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Sam"
          />
        </td>
        <td>
          <input
            className="form-control"
            onChange={[Function]}
            value="Atkins"
          />
        </td>
        <td>
          <button
            className="btn btn-success save-user-btn"
            onClick={[Function]}
          >
            Save
          </button>
          <button
            className="btn btn-warning"
            onClick={[Function]}
          >
            Clear
          </button>
        </td>
      </tr>
      <tr>
        <td>
          sam
        </td>
        <td>
          Sam
        </td>
        <td>
          Atkins
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          pparker
        </td>
        <td>
          Peter
        </td>
        <td>
          Parker
        </td>
        <td>
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="btn btn-danger delete-user-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
